### ISSUE TITLE
ishermitian/issymmetric should accept optional atol keyword parameter

### ISSUE THREAD

**Issue by GiggleLiu** [JuliaLang/LinearAlgebra.jl#558]
I was expecting the following code print two `true`s.
```
using LinearAlgebra
siz = 8
base_space = qr(randn(Float64, siz, siz)).Q
phases = rand(siz)
signs = Diagonal(phases)
A = base_space*signs*base_space'
println(A ≈ A')
println(ishermitian(A))
```
Instead, I get `true` and `false`.

The `ishermitian` turns out using `==`. Isn't it better to use `≈` or add an `atol` keyword parameter?
https://github.com/JuliaLang/julia/blob/4498d27c3c3e11effdfdfdce822d025079f01a0a/stdlib/LinearAlgebra/src/generic.jl#L971

**Comment by GiggleLiu** [comment]
I think this "feature" is not consistent with other julia methods such as `eigen`.
`eigen(A)` here will return **real** eigenvalues, which means the program take `A` as a hermitian matrix, although it has ~1e-16 error.

**Comment by KristofferC** [comment]
See the discussion in https://github.com/JuliaLang/LinearAlgebra.jl/issues/182.

**Comment by garrison** [comment]
I agree that the default behavior of `ishermitian` should not be changed (in line with previous discussion), but in my opinion an optional `atol` keyword argument would be a welcome feature.

**Comment by bjarthur** [comment]
simple enough to redefine `ishermitian` yourself if the julia devs have decided not to add a tolerance argument:

```
julia> import LinearAlgebra: ishermitian

julia> using LinearAlgebra

julia> function ishermitian(A::AbstractMatrix; isapprox=isapprox)
           indsm, indsn = axes(A)
           if indsm != indsn
               return false
           end
           for i = indsn, j = i:last(indsn)
               if !isapprox(A[i,j], adjoint(A[j,i]))
                   return false
               end
           end
           return true
       end
ishermitian (generic function with 14 methods)

julia> ishermitian([1 2; 2.00000000000001 1])
true
```

**Comment by timholy** [comment]
It's possibly less of a decision and more that no one has submitted a PR? But I suspect there is little point since lots of other routines will fail or produce surprising results unless it's exact.

**Comment by garrison** [comment]
> It's possibly less of a decision and more that no one has submitted a PR?

If this is the case, it could be clarified by having an open issue regarding this missing feature.  This could be accomplished most easily by opening and renaming this one, while perhaps adjusting the top comment (I no longer have the permission to do any of this), or alternatively by having a new issue elsewhere.

Suggested [edited] title: `ishermitian` should accept optional `atol` keyword parameter

**Comment by garrison** [comment]
> But I suspect there is little point since lots of other routines will fail or produce surprising results unless it's exact.

In my own personal experience of dealing with quantum mechanics numerically, the most common use case for this missing feature has been as a sanity check that a matrix I have constructed is actually Hermitian before treating it conceptually as such.  Without such a check, it is much easier to let a non-physical Hamiltonian or observable slip through.

Often, the matrix I have constructed will be non-Hermitian only by the tiniest amounts (e.g. because floating point addition is non-associative), and I'd like to perform a conceptual test of whether the matrix is "close enough" to Hermitian.

**Comment by timholy** [comment]
Reopened and renamed as suggested.

**Comment by bjarthur** [comment]
my use case is also as a sanity check to make sure that a numerically simulated matrix that should be symmetric actually is.

**Comment by bjarthur** [comment]
my re-definition above would be better if the added keyword argument were `isequal=isequal` and then folks like me and @garrison could use `isequal=isapprox`.  for everyone else nothing changes and completely backward compatible.

**Comment by bjarthur** [comment]
@garrison see https://github.com/JuliaLang/LinearAlgebra.jl/issues/737.  `isapprox(A,A')` suffices for me.  hadn't thought of doing that.

**Comment by garrison** [comment]
Oh, interesting find regarding JuliaLang/LinearAlgebra.jl#737.  I had been thinking recently that anything decided here should apply to `issymmetric` as well.

**Comment by oscardssmith** [comment]
`isapprox(A,A')` will probably be about 2x slower, but that's probably not a major problem for many use cases.

**Comment by jlapeyre** [comment]
I wrote a package a while back to try to approach specifying  measure of closeness in a uniform way: [IsApprox.jl](https://github.com/jlapeyre/IsApprox.jl). I included code for many functions/methods including `ishermitian`.
> Four subtypes of `AbstractApprox` are included: `Equal`, `Approx`, `EachApprox`, and `UpToPhase`.
IsApprox implements the interface at least partially for each of: isone, iszero, ishermitian, issymmetric, isreal, isinteger, istriu, istril, isbanded, isdiag, isposdef, ispossemidef, isunitary, isinvolution.
